addon/utils/decorator-macros.js

```diff
@@ -29,3 +29,20 @@ export function decoratedPropertyWithRequiredParams(fn) {
     return fn(...params);
   });
 }
+
+export function decoratedPropertyWithOptionalCallback(fn) {
+  return decoratorWithParams(function(target, key, desc, params) {
+    assert(`Cannot use '${fn.name}' on field '${key}' without parameters`, params.length !== 0);
+
+    if (typeof params[params.length - 1] === 'function') {
+      return fn(...params);
+    }
+
+    const value = extractValue(desc);
+    if (typeof value === 'function') {
```

Comment:
Could probably do this check eagerly in the `assert` below so that it gets stripped in production